[DOC] Adds developer guideline start (#446)
ME-ICA/tedana: docs/developing.rst

Developer Guidelines
====================

> Git Workflow

Reviewer: I really think this should go back to the CONTRIBUTING...

Author: The `git` workflow or the whole page?

NOTE: if you don't understand the above steps, consider using a git tutorial and a practice repository until you get the hang of it.
Feel free also to ask questions and use a ``git`` client.
Several popular ones include

- GitKraken_
- `GitHub Desktop`_
- SourceTree_

Adding and Modifying Tests
--------------------------
Testing is an important component of development.
For simplicity, we have migrated all tests to ``pytest``.
There are two basic kinds of tests:
unit and integration tests.
Unit tests focus on testing individual functions,
whereas integration tests focus on making sure that the whole workflow
runs correctly.

For unit tests,
we try to keep tests on the same module grouped into one file.
Make sure the function you're testing is imported,
then write your test.
Good tests will make sure that edge cases are accounted for as well as
common cases.
You may also use ``pytest.raises`` to ensure that errors are thrown for
invalid inputs to a function.

For integration tests,
make a ``tar.gz`` file which will unzip to be a single directory
containing all of the files you'd like to run a workflow on.
Run the workflow with a known-working version, and put the outputs into a

> text file inside ``TEDANADIR/tedana/tests/data/``.

Reviewer: Same concern about `TEDANADIR`, here.

====================
Developer Guidelines
====================

This webpage is intended to guide users through making making changes to
``tedana``'s codebase, in particular working with tests.
The worked example also offers some guidelines on approaching testing when
adding new functions.

Adding and Modifying Tests
==========================
Testing is an important component of development.
For simplicity, we have migrated all tests to ``pytest``.
There are two basic kinds of tests: unit and integration tests.
Unit tests focus on testing individual functions, whereas integration tests focus on making sure
that the whole workflow runs correctly.

> For unit tests, we try to keep tests on the same module grouped into one file.

Reviewer: ```suggestion
For unit tests, we try to keep tests from the same module grouped into one file.
```

Reviewer: I'd add a subheading to this to specify that it's unit testing (as opposed to other cases outlined below)

Make sure the function you're testing is imported, then write your test.
Good tests will make sure that edge cases are accounted for as well as common cases.
You may also use ``pytest.raises`` to ensure that errors are thrown for invalid inputs to a
function.

> For integration tests, make a ``tar.gz`` file which will unzip to be only the files you'd like to

Reviewer: I would couch this -- it's going to be fairly uncommon at this point. I might say something like:

> In rare cases, you may need to add a new integration test. This means adding a new kind of dataset that tedana will be explicitly tested to support. As always, please open an issue first if you think that you have such a dataset ! Then, after making sure you have appropriate permission to share it, follow these steps to add it to the tedana repository.

Reviewer: Also, this should be a new section !

Author: Addressed in 6a6a71f and 13e3daa

run a workflow on.
You can do this with the following, which would make an archive ``my_data.tar.gz``:

.. code-block:: bash

    tar czf my_data.tar.gz my_data/*.nii.gz

Run the workflow with a known-working version, and put the outputs into a text file inside

> ``$TEDANADIR/tedana/tests/data/``,with ``TEDANADIR`` the local ``tedana repository``.

Reviewer 2: ```suggestion
``$TEDANADIR/tedana/tests/data/``, where ``TEDANADIR`` is your local ``tedana repository``.
```

> You can follow the model our `five echo set`_, which has the following steps:

Reviewer 2: ```suggestion
To write the test function you can follow the model of our `five echo set`_, which takes the following steps:
```

1. Check if a pytest user is skipping integration, skip if so
#. Use ``download_test_data`` to retrieve the test data from OSF
#. Run a workflow
#. Use ``resources_filename`` and ``check_integration_outputs`` to compare your expected output to
   actual output.

If you need to upload new data, you will need to contact the maintainers and ask them to either add

> it or give you permission to add it.

Reviewer 2: ```suggestion
it to `OSF`_ or give you permission to add it.
```

And then can you add the `OSF`_ link to the links at the bottom?

Once you've tested your integration test locally and it is working, you will need to add it to the
CircleCI config and the ``Makefile``.
Following the model of the three-echo and five-echo sets, define a name for your integration test
and on an indented line below put

.. code-block:: bash

    @py.test --cov-append --cov-report term-missing --cov=tedana -k TEST

with ``TEST`` your test function's name.
This call basically adds code coverage reports to account for the new test, and runs the actual
test in addition.
Using the five-echo set as a template, you should then edit ``.circlec/config.yml`` to add your
test, calling the same name you define in the ``Makefile``.

> If you need to take a look at a failed test on CircleCI rather than locally, you can use the

Reviewer: New section here, please.

Author: Done in c720db1

following block to retrieve artifacts (see CircleCI documentation here_)

.. code-block:: bash

    export CIRCLE_TOKEN=':your_token'

    curl https://circleci.com/api/v1.1/project/:vcs-type/:username/:project/$build_number/artifacts?circle-token=$CIRCLE_TOKEN \
       | grep -o 'https://[^"]*' \
       | sed -e "s/$/?circle-token=$CIRCLE_TOKEN/" \
       | wget -v -i -

To get a CircleCI token, follow the instructions for `getting one`_.
You cannot do this unless you are part of the ME-ICA/tedana organization.
If you don't want all of the artifacts, you can go to the test details and use the browser to
manually select the files you would like.

> Worked Example

Reviewer: Can we add pictures, here ??

Author: What specifically would you like pictures for?

sure you don't forget to create it!)
Since the function lives in ``io.py``, its unit test should go into ``test_io.py``.
The job of this test is exclusively to tell if the function we wrote does what it claims to do
without errors.
So, we define a new function in ``test_io.py`` that looks something like this:

.. code-block:: python

    def test_say_hello():
        # run the function
        say_hello()
        # test the function
        assert op.exists('hello_world.txt')
        # clean up
        os.remove('hello_world.txt')

We should see that our unit test is successful via

.. code-block:: bash

    pytest $TEDANADIR/tedana/tests/test_io.py -k test_say_hello
    
If not, we should continue editing the function until it passes our test.
Let's suppose that suddenly, you realize that what would be even more useful is a function that
takes an argument, ``place``, so that the output filename is actually ``hello_PLACE``, with
``PLACE`` the value passed and ``'world'`` as the default value.
We merge any changes from the upstream master branch into our branch via

.. code-block:: bash

>     git checkout feature/say_hello  # unless you're already there

Reviewer: It won't hurt even if you are, right ?

Author: That's correct, I could amend the wording to make that more clear if you like.

    git fetch upstream master
    git merge upstream/master

and then begin work on our test.
We need to our unit test to be more complete, so we update it to look more like the following,
adding several cases to make sure our function is robust to the name supplied:

.. code-block:: python

    def test_say_hello():
        # prefix of all files to be checked
        prefix = 'hello_'
        # suffix of all files to be checked
        suffix  = '.txt'
        # run the function with several cases
        for x in ['world', 'solar system', 'galaxy', 'universe']:
            # current test name
            outname = prefix + x + suffix
            # call the function
            say_hello(x)
            # test the function
            assert op.exists(outname)
            # clean up from this call
            os.remove(outname)

Once that test is passing, we may need to adjust the integration test.
Our program creates a file, ``hello_world.txt``, which the older version would not have produced.
Therefore, we need to add the file to ``$TEDANADIR/tedana/tests/data/tedana_outputs.txt`` and its

> counterpart, R2-D2-- uh, we mean, ``tedana_outputs_verbose.txt``.

Reviewer: :laughing:

With that edit complete, we can run the full ``pytest`` suite via

.. code-block:: bash

    pytest $TEDANADIR/tedana/tests

Once that filename is added, all of the tests should be passing and we should open a PR to have our
change reviewed.

From here, others working on the project may request changes and we'll have to make sure that our
tests are kept up to date with any changes made as we did before updating the unit test.
For example, if a new parameter is added, ``greeting``, with a default of ``hello``, we'll need to
adjust the unit test.
However, since this doesn't change the typical workflow of ``tedana``, there's no need to change
the integration test; we're still matching the original filename.
Once we are happy with the changes and some members of ``tedana`` have approved the changes, our
changes will be merged!

We should then do the following cleanup with our git repository:

.. code-block:: bash

    git checkout master
    git fetch upstream master
    git merge upstream/master
    git branch -d feature/say_hello
    git push --delete origin feature/say_hello

and we're good to go!

> Monthly Developer Calls

Reviewer: I feel like this is sort of buried, here. Can we put it at the top of the page ?

Author: Sure! I wasn't sure what priority this has versus other sections.

Author: Fixed in 7102ff0

=======================
We run monthly developer calls via Zoom.
You can see the schedule via the tedana `google calendar`_.

Everyone is welcome.

> We look forward to meeting you there :hibiscus:

Reviewer 2: ```suggestion
We look forward to meeting you there!
```

The emoji aren't supported on .rst, unfortunately 😢

Author: Even worse, there is no sad hibiscus emoji for me to respond with.

====================
Developer Guidelines
====================

This webpage is intended to guide users through making making changes to
``tedana``'s codebase, in particular working with tests.
The worked example also offers some guidelines on approaching testing when
adding new functions.

> 

Reviewer 3: I would add a link to your contributing guide somewhere up here

sure you don't forget to create it!)
Since the function lives in ``io.py``, its unit test should go into ``test_io.py``.
The job of this test is exclusively to tell if the function we wrote does what it claims to do
without errors.
So, we define a new function in ``test_io.py`` that looks something like this:

.. code-block:: python

    def test_say_hello():
        # run the function
        say_hello()
        # test the function
        assert op.exists('hello_world.txt')
        # clean up
        os.remove('hello_world.txt')

We should see that our unit test is successful via

.. code-block:: bash

    pytest $TEDANADIR/tedana/tests/test_io.py -k test_say_hello
    
If not, we should continue editing the function until it passes our test.
Let's suppose that suddenly, you realize that what would be even more useful is a function that
takes an argument, ``place``, so that the output filename is actually ``hello_PLACE``, with
``PLACE`` the value passed and ``'world'`` as the default value.
We merge any changes from the upstream master branch into our branch via

.. code-block:: bash

>     git checkout feature/say_hello  # unless you're already there

Reviewer 2: ```suggestion
    git checkout feature/say_hello
```

Author: Hm, I did this because @emdupre had pointed that it's unclear that you only have to do this if you've left the branch. Do you think this adds to the confusion instead of clarifying?

====================
Developer Guidelines
====================

Adding and Modifying Tests
==========================
Testing is an important component of development.
For simplicity, we have migrated all tests to ``pytest``.
There are two basic kinds of tests:
unit and integration tests.
Unit tests focus on testing individual functions,
whereas integration tests focus on making sure that the whole workflow
runs correctly.

For unit tests,
we try to keep tests on the same module grouped into one file.
Make sure the function you're testing is imported,
then write your test.
Good tests will make sure that edge cases are accounted for as well as
common cases.
You may also use ``pytest.raises`` to ensure that errors are thrown for
invalid inputs to a function.

For integration tests,
make a ``tar.gz`` file which will unzip to be a single directory

> containing all of the files you'd like to run a workflow on.

Author: I meant to type "archive," whoops. Addressed in a63ccfe

and ask them to either add it or give you permission to add it.
Once you've tested your integration test locally and it is working,
you will need to add it to the CircleCI config and the ``Makefile``.
Following the model of the three-echo and five-echo sets,
define a name for your integration test and on an indented line below put 

.. code-block:: none

    @py.test --cov-append --cov-report term-missing --cov=tedana -k TEST
with ``TEST`` your test function's name. 
This call basically adds code coverage reports to account for the new test,
and runs the actual test in addition.
Using the five-echo set as a template,
you should then edit ``.circlec/config.yml`` to add your test,
calling the same name you define in the ``Makefile``.

If you need to take a look at a failed test on CircleCI rather than
locally, you can use the following block to retrieve artifacts
(see CircleCI documentation here_)

.. code-block:: none
    export CIRCLE_TOKEN=':your_token'

    curl https://circleci.com/api/v1.1/project/:vcs-type/:username/:project/$build_number/artifacts?circle-token=$CIRCLE_TOKEN \
       | grep -o 'https://[^"]*' \
       | sed -e "s/$/?circle-token=$CIRCLE_TOKEN/" \

>        | wget -v -i -

Author: Addressed in d722317
